### PR TITLE
Qwen3ASR: double-buffer asyncEval greedy decode

### LINKED ISSUE #229: Qwen3ASR: reduce greedy decode host/GPU sync overhead
## Context

I come from an LLM/inference background and have been especially interested in improving local Mac inference efficiency. While integrating `Qwen3ASRModel` into a downstream macOS transcription app, I noticed that hot ASR inference on Apple Silicon was leaving noticeable GPU headroom during greedy decode. Activity Monitor showed GPU utilization around 50-60%, and the wall-clock profile looked consistent with a per-token host/GPU synchronization bubble rather than audio decoding.

This made me look at the path through the lens of LLM serving systems like vLLM and SGLang. I am not trying to port their full scheduler/batching stack here, but the same general idea applies: avoid unnecessary host synchronization and overlap CPU bookkeeping with already-enqueued device work where possible.

The current greedy path eventually chooses each token with:

```swift
argMax(logits, axis: -1).squeezed().item(Int32.self)
```

That `.item(Int32.self)` forces the GPU to finish the current decoder step and copy a scalar token id back to the CPU before Swift can build and enqueue the next step.

## Hypothesis

For the default greedy path, the next-token MLXArray can stay on-device long enough to stage the following decoder step first. Using `asyncEval` to kick off the next argmax/cache update before synchronizing the current token should overlap some CPU bookkeeping with the next GPU step.

This is not expected to be a large vLLM-style batching win, because the 1.7B 8-bit decoder is still mostly weight-bandwidth bound. But it should remove a small, repeatable host sync bubble without changing public API or non-greedy decoding behavior.

## Local experiment

I implemented a local branch that:

- routes only the default greedy path (`repetitionPenalty == 1`, `noRepeatNgramSize == 0`, `temperature == 0`) through a double-buffered `asyncEval` loop
- leaves the CPU-side repetition/no-repeat/sampling path unchanged
- keeps argmax dtype aligned with the legacy `Int32` token input path
- preserves the legacy append-then-EOS-break behavior

## Local benchmark

Downstream benchmark harness:

- model: `aufklarer/Qwen3-ASR-1.7B-MLX-8bit`
- audio: local lecture `.m4a`
- chunks: 4 x 15s = 60s audio per iteration
- iterations: 5, hot mean over iter 2-5
- runner: Xcode test path from a macOS app, so MLX metallib is available

| run | branch | hot wall | RTF | relative speed |
| --- | --- | ---: | ---: | ---: |
| baseline | `main` | 7.406s | 0.1234 | 1.000x |
| optimized | asyncEval branch | 6.838s | 0.1140 | **1.083x** |
| delta | — | **-0.568s** | **-0.0094** | **+8.3%** |

Per-chunk speedups were 1.075x-1.087x. This is a modest but clear improvement for a tiny API-neutral change.

Correctness check: 4 chunks produced matching text signatures across baseline and optimized runs in the same process lineage, and each branch was deterministic across 5 hot iterations. I did see one cross-process MLX nondeterminism case while investigating, so I avoided over-claiming absolute process-to-process bit identity.

## Future optimization space

This PR is meant as a small, low-risk first step rather than the end of the performance work. From an LLM serving perspective, the larger wins would likely come from infrastructure-level changes that better match ideas from vLLM / SGLang:

- **Static / continuous chunk batching**: offline ASR already has independent chunks, so batching multiple chunks through the decoder could amortize weight loads across several token streams. This is the most likely path to multi-x throughput improvements.
- **Speculative decoding**: a smaller Qwen3-ASR model could draft several tokens, then the 1.7B model could verify them in a short batched forward pass. ASR may be a good fit because the conditional entropy is relatively low.
- **Prefix / prompt KV reuse**: the repeated chat/template prefix could be cached across chunks for a smaller but composable win.

I kept the current PR deliberately narrow because it preserves public API and non-greedy behavior, and because it gives a measurable improvement without introducing a scheduler or multi-model state management. I would be very happy to discuss whether any of the larger infra directions fit the maintainers' roadmap.

## Testing caveat

Running `swift test` directly in this repo compiles, but MLX-backed tests on my machine hit:

```text
MLX error: Failed to load the default metallib. library not found
```

So the performance/correctness validation above was done through the downstream Xcode app test bundle where the MLX metallib is present. I am happy to adapt the benchmark or add a smaller repo-local regression test if you prefer a different validation shape.

## PR

@ivan-digital I have a small PR ready with the code change and would love feedback on whether this is an acceptable direction for Qwen3-ASR greedy decoding.

### PR DESCRIPTION
**Pull request by hhh2210**
### Problem

Qwen3-ASR greedy decoding currently synchronizes every generated token back to the CPU before the next decoder step can be staged:

```swift
argMax(logits, axis: -1).squeezed().item(Int32.self)
```

On a downstream macOS transcription app, this showed up as hot ASR runs leaving visible GPU headroom during decode (roughly 50-60% GPU utilization), with a profile consistent with a small host/GPU synchronization bubble per token.

### Background / motivation

My background is mostly around LLMs and inference/runtime behavior, and I have been particularly interested in making local Mac inference more efficient. This patch borrows the high-level intuition from systems like vLLM and SGLang: keep device work queued and avoid synchronizing to the host earlier than necessary.

This PR is intentionally much smaller than batching/speculative decoding. It only targets the default greedy Qwen3-ASR path and tries to remove one avoidable synchronization bubble without changing public API.

### Cause

The default greedy path only needs the scalar token id on CPU for EOS checking and appending to the returned token list. The next decoder step itself can be staged from an on-device `MLXArray` token result first, then the current token can be synchronized afterward.

### Fix

This PR adds a greedy-only double-buffered decode loop for `Qwen3ASRModel.generateText`:

- routes only the default greedy options through the new path (`repetitionPenalty == 1`, `noRepeatNgramSize == 0`, `temperature == 0`)
- stages the next token embedding / decoder forward from the lazy argmax result before synchronizing the current token id
- calls `asyncEval(nextTokenArr, newCache)` to kick GPU work before CPU bookkeeping
- keeps `argMax` dtype aligned with the legacy `Int32` token input path
- preserves the legacy append-then-EOS-break behavior
- leaves repetition penalty, no-repeat n-gram, and temperature sampling on the existing CPU-side path

### Local testing

Validated in a downstream Xcode app test bundle where MLX's `default.metallib` is available:

- model: `aufklarer/Qwen3-ASR-1.7B-MLX-8bit`
- audio: local lecture `.m4a`
- chunks: 4 x 15s = 60s audio per iteration
- iterations: 5, hot mean over iter 2-5

| run | branch | hot wall | RTF | relative speed |
| --- | --- | ---: | ---: | ---: |
| baseline | `main` | 7.406s | 0.1234 | 1.000x |
| optimized | this branch | 6.838s | 0.1140 | **1.083x** |
| delta | — | **-0.568s** | **-0.0094** | **+8.3%** |

Per-chunk speedups were 1.075x-1.087x.

Correctness check: the four benchmark chunks produced matching text signatures between baseline and this branch in the same process lineage, and were deterministic across 5 hot iterations.

### Repo-local validation caveat

I also tried `swift test` directly in this repo. Compilation succeeds, but MLX-backed tests fail locally with:

```text
MLX error: Failed to load the default metallib. library not found
```

That appears to be a local SwiftPM/MLX metallib availability issue, not specific to this PR. I am happy to add or adjust a repo-local regression/benchmark if there is a preferred way to run MLX inference tests here.

Fixes #229

Love to discuss further with the code owner. The win is modest rather than dramatic, but the local testing suggests it is repeatable and comes from a small, API-neutral change to the default greedy Qwen3-ASR path.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by ivan-digital** [comment]
Thanks for the patch! Verified bit-exact greedy and ~5% wall-clock speedup on M-series Mac with the 1.7B-8bit model on a 71 s multilingual clip — a bit less than the GPU-headroom numbers would suggest, but real and reproducible. Following up with a small PR adding a snapshot test (locks in the bit-exactness invariant, including the uint32→int32 argMax cast) plus a doc note in `docs/inference/qwen3-asr-inference.md`.
